### PR TITLE
[REVIEW] Avoid using numpy via cimport to prevent ABI issues

### PR DESCRIPTION
**Pull request by kkraus14**
cc @dantegd and @thomcom for visibility

### CODE REVIEW

**Comment by kkraus14** [comment]
Needs #821 cherry picked into here to resolve the style issues from a new version of flake8.

File: python/cudf/bindings/cudf_cpp.pxd [hunk 1]
```diff
@@ -6,7 +6,8 @@
 # cython: language_level = 3
 
 from libcpp cimport bool
-from numpy cimport uint8_t, int64_t, int32_t, int16_t, int8_t
+#from numpy cimport uint8_t, int64_t, int32_t, int16_t, int8_t
```
Comment (thomcom) [inline python/cudf/bindings/cudf_cpp.pxd]: Delete this, git will remember for us.

Comment (kkraus14) [reply python/cudf/bindings/cudf_cpp.pxd]:
```suggestion
```

**Comment by kkraus14** [comment]
rerun tests

**Comment by kkraus14** [comment]
rerun tests

**Comment by kkraus14** [comment]
rerun tests

File: python/setup.py [hunk 2]
```diff
@@ -3,7 +3,7 @@
 from setuptools import setup, find_packages
 from setuptools.extension import Extension
 from Cython.Build import cythonize
-import numpy
+# import numpy
```
Comment (mt-jones) [inline python/setup.py]: Recommend removing comments.

File: python/setup.py [hunk 3]
```diff
@@ -14,17 +14,12 @@
     'cython'
 ]
 
-try:
-    numpy_include = numpy.get_include()
-except AttributeError:
-    numpy_include = numpy.get_numpy_include()
-
 cython_files = ['cudf/bindings/*.pyx']
```
Comment (mt-jones) [inline python/setup.py]: I'm not sure a wild-card is the best way to aggregate the appropriate Cython files.

File: python/cudf/bindings/join.pyx [hunk 4]
```diff
@@ -13,16 +13,17 @@ from .cudf_cpp import *
...
 
 from librmm_cffi import librmm as rmm
 
 from libc.stdint cimport uintptr_t
 from libc.stdlib cimport calloc, malloc, free
 
+cimport cython
 
 
+@cython.boundscheck(False)
```
Comment (mt-jones) [inline python/cudf/bindings/join.pyx]: Why are we disabling bounds-checking?

File: python/cudf/bindings/join.pyx [hunk 5]
```diff
@@ -88,10 +89,10 @@ cpdef join(col_lhs, col_rhs, on, how, method='sort'):
         if how == 'left':
             result = gdf_left_join(list_lhs,
                 col_lhs_len,
-                <int*>left_idx.data,
+                &left_idx[0],
```
Comment (mt-jones) [inline python/cudf/bindings/join.pyx]: Will `gdf_*_join` perform the proper bounds-checks with this pointer and the length(s)?

File: python/cudf/bindings/join.pyx (hunk 4 above)
Comment (dantegd) [reply python/cudf/bindings/join.pyx]: To disable a warning that appeared when this PR changed the way we created NumPy arrays.

File: python/cudf/bindings/join.pyx (hunk 5 above)
Comment (dantegd) [reply python/cudf/bindings/join.pyx]: AFAIK its up to the C++ side, the cython boundscheck is just for the cython side of the code (i.e. if you were doing the computations with cython code), C++ can still out of bound with this side of bounds check. Might create a quick test of that in a little bit if this PR is still open by then...

File: python/setup.py (hunk 3 above)
Comment (dantegd) [reply python/setup.py]: It is the standard way, and out of the scope of this PR

Comment (dantegd) [reply python/setup.py]: (not totally disagreeing with you though, not doing it that way in cuML to be fair, would like to talk outside of the PR)

Comment (kkraus14) [reply python/setup.py]: Out of scope of this PR.

File: python/setup.py (hunk 2 above)
Comment (kkraus14) [reply python/setup.py]:
```suggestion
```

File: python/setup.py (hunk 3 above)
Comment (mt-jones) [reply python/setup.py]: I would tend to disagree that it's out of scope of this PR. If you're making changes to the `setup.py` file in a hot-fix, then it makes sense to address relatively small adjustments for consistency with respect to other elements of our software ecosystem.

A wild-card will grab any `.pyx` file extension upon install, which is a very easy place for errors to occur. Specifying the file list avoids potential errors, and makes this more robust/intentional.

If we want to put it into another PR, that's fine, but I don't know the time we'll have reason to touch this file.

File: python/cudf/bindings/join.pyx (hunk 4 above)
Comment (mt-jones) [reply python/cudf/bindings/join.pyx]: That makes sense. Thanks for clarifying.

File: python/cudf/bindings/join.pyx (hunk 5 above)
Comment (mt-jones) [reply python/cudf/bindings/join.pyx]: It makes sense to me to have bounds checks for both Cython and C++. These checks create more visibility into errors that crop up when calls to this binding are made. Not sure if this belongs here, though.

Comment (kkraus14) [reply python/cudf/bindings/join.pyx]: Given it's a warning and we do pretty extensive bounds checking in the functions an end user would call above this I'm going to close this for now, but we can revisit if it's an issue later.

File: python/setup.py (hunk 3 above)
Comment (kkraus14) [reply python/setup.py]: Given we merged this already and we wanted to do something quick and dirty for a hotfix, lets restart this discussion for 0.6 along with further build improvements across both python and CMake.

**Comment by thomcom** [comment]
Into `master`?

**Comment by kkraus14** [comment]
> Into `master`?

Yes this is a hotfix because it broke ABI compatibility with numpy between versions / distributions.

**Pull request merged**